File: Aulas/Aula5/Heranca-Multipla.java
// Interface Animal
abstract class Animal {
    protected String nome;

    public Animal(String nome) {
        this.nome = nome;
    }

    abstract void emitirSom();
}

// Classe Mamifero que estende Animal
class Mamifero extends Animal {
    public Mamifero(String nome) {
        super(nome);
    }

    public void amamentar() {
        System.out.println(nome + " está amamentando.");
    }

    @Override
    void emitirSom() {
        // Método abstrato pode ser implementado nas subclasses
    }
}

// Interface Ave com o comportamento de voar
interface Ave {
    void voar();
}

// Classe Morcego que estende Mamifero e implementa a interface Ave
class Morcego extends Mamifero implements Ave {
    public Morcego(String nome) {
        super(nome);
    }

    @Override
    public void emitirSom() {
        System.out.println("Morcego fazendo ruídos noturnos.");
    }

    @Override
    public void voar() {
        System.out.println(nome + " está voando.");
    }

    public static void main(String[] args) {
        Morcego morcego = new Morcego("Batemane");
        morcego.emitirSom();    // Método da classe Morcego
        morcego.amamentar();    // Método da classe Mamifero
        morcego.voar();         // Método da interface Ave
    }
}

File: Aulas/Aula5/Heranca-Multipla.py
class Animal:
    def __init__(self, nome):
        self.nome = nome

    def emitir_som(self):
        pass  # Método a ser implementado nas subclasses

class Mamifero(Animal):
    def __init__(self, nome):
        super().__init__(nome)

    def amamentar(self):
        print(f"{self.nome} está amamentando.")

class Ave(Animal):
    def __init__(self, nome):
        super().__init__(nome)

    def voar(self):
        print(f"{self.nome} está voando.")

class Morcego(Mamifero, Ave):
    def __init__(self, nome):
        super().__init__(nome)

    def emitir_som(self):
        print("Morcego fazendo ruídos noturnos.")

# Exemplo de uso
morcego = Morcego("Batemane")
morcego.emitir_som()  # Método da classe Animal
morcego.amamentar()   # Método da classe Mamifero
morcego.voar()        # Método da classe Ave
